File: 0015 3Sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        n = len(nums)
        for i in range(n - 2):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            lo = i + 1
            hi = n - 1
            total = -nums[i]
            while (lo < hi):
                if nums[lo] + nums[hi] == total:
                    res.append([nums[i], nums[lo], nums[hi]])
                    while lo < hi and nums[lo + 1] == nums[lo]:
                        lo += 1
                    while lo < hi and nums[hi - 1] == nums[hi]:
                        hi -= 1
                    lo += 1
                    hi -= 1
                elif nums[lo] + nums[hi] > total:
                    hi -= 1
                else:
                    lo += 1
        return res 

File: 0015 3Sum/3Sum.java
class 3Sum {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        Arrays.sort(nums); // sort the array
        for (int i = 0; i < nums.length - 2; i++) {
            if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {   
                int lo = i + 1;  // two pointers to scan the array
                int hi = nums.length - 1;
                int sum = 0 - nums[i];
                while (lo < hi) {
                    if (nums[lo] + nums[hi] == sum) {
                        res.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
                        while (lo < hi && nums[lo] == nums[lo + 1]) {  // get rid of duplicated integers
                            lo++;
                        }
                        while (lo < hi && nums[hi] == nums[hi - 1]) {
                            hi--;
                        }
                        lo++;
                        hi--;
                    }
                    else if (nums[lo] + nums[hi] < sum) {
                        lo++;
                    }
                    else {
                        hi--;
                    }
                }
            }
        }
        return res;
    }
}